week06/kimsehyun.md

```diff
@@ -0,0 +1,533 @@
+	  - '8000:80'
+	volumes:
+	  # 서버도 결국 php에 필요한 소스코드를 필요로 하기 때문에, bind mount로 마운트해준다.
+	  # - ./src:/var/www/html
+	  
+	  # docker nginx 이미지 공식 페이지에서 다음과 같이 제시
+	  # 로컬에다 nginx > nginx.conf 파일을 따로 만들어야 함(강의 자료에 첨부)
+
+      # 잘못된 예시 (에러 발생)
+	  # - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
+	  # 올바른 예시
+      # - ./nginx/nginx.conf:/etc/nginx/conf.d/default.conf:ro
+	
+	  # server 서비스는 php와 mysql 서비스가 실행되어야만 작동하기 떄문(의존하기 때문)
+      # server 서비스를 실행하면, php와 mysql 서비스도 자동으로 실행하게 한다.  
+	depends_on:
+	  - php
+	  - mysql
+
+  php:
+    build:
+	  context: .
+	  dockerfile: dockerfiles/php.dockerfile
+	# 배포 단계에서는 bind mount를 따로 설정하지 않는다
+    volumes:
+      - ./src:/var/www/html:delegated
+    ports:
+      # 도커 php base image 공식 소스코드를 보면 default 포트가 9000
+      # 하지만 사실 해줄 필요가 없다.
+      # - '3000:9000'
+
+  mysql:
+    image: mysql:5.7
+    # mysql DB를 사용하려면 계정 정보가 있어야 하고, 
+    # 이런 개인정보는 환경변수로 저장하는 게 보안을 위해 좋다.
+    env_file:
+      - ./env/mysql.env
+  
+  composer:
+    build:
+      context: ./dockerfiles
+      dockerfile: composer.dockerfile
+    volumes:
+      - ./src:/var/www/html
+
+  artisan:
+    build:
+      context: .
+      dockerfile: ./dockerfiles/php.dockerfile
+    volumes:
+      - ./src:/var/www/html
+    entrypoint: ["php", "/var/www/html/artisan"]
+
+  npm:
+    image: node:14
+    working_dir: /var/www/html
+    entrypoint: ["npm"]
+    volumes:
+      - ./src:/var/www/html
+
+```
+
+
+---
+
+
+### Mac Apple Silicon에서 실행 에러
+
+다음과 같은 에러 발생: `no matching manifest for linux/arm64/v8 in the manifest list entries`
+
+
+![[2024-01-18_21-24-41.png]]
+
+해결책: docker compose 파일에서 mysql 설정에 이렇게 platform 정보를 추가해준다.
+
+```yaml
+  mysql:
+    image: mysql:5.7
+    platform: linux/amd64
+    env_file:
+      - ./env/mysql.env
+```
+
+
+
+### PHP 버전 때문에 발생하는 syntax 에러
+```
+Parse error: syntax error, unexpected '|', expecting variable (T_VARIABLE) in /var/www/html/vendor/nunomaduro/termwind/src/Functions.php on line 17
+```
+
+composer와 php 버전 차이 때문에 발생한 듯
+https://stackoverflow.com/questions/69072898/parse-error-syntax-error-unexpected-expecting-variable-t-variable-symfo
+
+> This character '|' is used only in php version >= 8.0 You can update the php version
+
+
+dockerfile을 직접 새로 하나 만들어서 docker compose에 지정하든가,
+아니면 기존 dockerfile를 사용하되 entrypoint를 추가하든가.
+
+base image에 대해서만 docker compose에 working_dir와 entrypoint를 지정하는 경우가 대부분.
+
+### bind mount
+혼자 develop하기엔 파일을 바로바로 미러링할 수 있어서 좋지만, deploy할 때는 지양해야 한다.
+
+
+
+
+---
+
+### 부록: delegated 옵션으로 docker volume 설정하기
+
+볼륨의 파일 전송 상태를 설정하는 법은 delegated 포함해서 총 세 가지가 있다.
+
+- `default` or `consistent`
+- `delegated`
+- `cached`
+
+"delegated" 옵션을 직역하자면 파일 전송의 권한을 컨테이너에게 위임한다는 뜻으로, 컨테이너가 자기 폴더 안에 있는 파일을 로컬 호스트 머신의 폴더에 업데이트할 때 호스트의 폴더를 read-only로 "꼼짝 못하게 묶어둔다"고 이해하면 된다. 
+
+한 가지 의문점은,  스택오버플로우(https://stackoverflow.com/questions/43844639/how-do-i-add-cached-or-delegated-into-a-docker-compose-yml-volumes-list)에선 컨테이너의 파일 업데이트가 빈번하게 일어날 때 delegated 옵션을 쓴다고 나와있는데 왜 강의 설명은 반대인지 모르겠다.
```

Comment:
#63 
파일업데이트가 빈번하지않을때 delegated 옵션쓴다고했다 

빈번할때인가 않을때인가 확인필요

---

\>  컨테이너 내부에서 파일들이 생성되기는 하지만, 해당 파일들이 호스트 머신과 자주 동기화될 필요가 없기 때문에 성능이 최적화된다는 의미인듯합니다.
- 컨테이너 내부의 파일시스템 업데이트는 빈번하게 일어남
- 컨테이너 외부(호스트 머신)로의 파일 시스템 업데이트를 줄여 성능 최적화

114강 PHP 컨테이너 추가 6:18~7:35
```
이것은 결국 컨테이너가 일부 데이터를 기록해야 하는 경우에
그에 대한 결과를 호스트 머신에 즉시 반영하지 않고
그 대신, 배치(batch)로 기본 처리함으로써
그 성능이 약간 더 나아집니다. (역자주: 안정성은 떨어지지만, 속도가 향상됨. 자세한 것은 구글링!)
...
이 컨테이너는 이 폴더에 자주 기록하지 않아야 하기 때문에...
또는 올바르게 기록을 하더라도
호스트 머신에 그 결과가 즉시 반영될 필요가 없습니다.
호스트 머신 상에서 즉시 액세스해야 하는 항목을 기록하지는 않으니까요.

(참고차 영어 원문)
in the end that means that if the containers
should write some data there,
it's not instantly reflected back to the host machine,
instead it is basically processed in batches, you could say,
and therefore performance is a bit better.

So adding delegated here improves performance
since this container shouldn't write frequently
to this folder,
or even if it does the things it does right
don't need to be reflected back
to the host machine instantly.
Because it won't write anything which
we need to access instantly on the host machine.
```